File: SignalShareExtension/SAEScreenLockViewController.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "SAEScreenLockViewController.h"
#import <SignalMessaging/SignalMessaging-Swift.h>
#import <SignalServiceKit/AppContext.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>
#import <SignalUI/SignalUI-Swift.h>
#import <SignalUI/Theme.h>

NS_ASSUME_NONNULL_BEGIN

@interface SAEScreenLockViewController () <ScreenLockViewDelegate>

@property (nonatomic, readonly, weak) id<ShareViewDelegate> shareViewDelegate;

@property (nonatomic) BOOL hasShownAuthUIOnce;

@property (nonatomic) BOOL isShowingAuthUI;

@end

#pragma mark -

@implementation SAEScreenLockViewController

- (instancetype)initWithShareViewDelegate:(id<ShareViewDelegate>)shareViewDelegate
{
    self = [super init];
    if (!self) {
        return self;
    }

    _shareViewDelegate = shareViewDelegate;

    self.delegate = self;

    return self;
}

- (void)loadView
{
    [super loadView];

    self.view.backgroundColor = Theme.launchScreenBackgroundColor;

    self.title = OWSLocalizedString(@"SHARE_EXTENSION_VIEW_TITLE", @"Title for the 'share extension' view.");

    self.navigationItem.leftBarButtonItem =
        [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemStop
                                                      target:self
                                                      action:@selector(dismissPressed:)];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    [self ensureUI];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];

    [self ensureUI];

    // Auto-show the auth UI f
    if (!self.hasShownAuthUIOnce) {
        self.hasShownAuthUIOnce = YES;

        [self tryToPresentAuthUIToUnlockScreenLock];
    }
}

- (void)dealloc
{
    // Surface memory leaks by logging the deallocation of view controllers.
    OWSLogVerbose(@"Dealloc: %@", self.class);
}

- (void)tryToPresentAuthUIToUnlockScreenLock
{
    OWSAssertIsOnMainThread();

    if (self.isShowingAuthUI) {
        // We're already showing the auth UI; abort.
        return;
    }
    OWSLogInfo(@"try to unlock screen lock");

    self.isShowingAuthUI = YES;

    [OWSScreenLock.shared
        tryToUnlockScreenLockWithSuccess:^{
            OWSAssertIsOnMainThread();

            OWSLogInfo(@"unlock screen lock succeeded.");

            self.isShowingAuthUI = NO;

            [self.shareViewDelegate shareViewWasUnlocked];
        }
        failure:^(NSError *error) {
            OWSAssertIsOnMainThread();

            OWSLogInfo(@"unlock screen lock failed.");

            self.isShowingAuthUI = NO;

            [self ensureUI];

            [self showScreenLockFailureAlertWithMessage:error.userErrorDescription];
        }
        unexpectedFailure:^(NSError *error) {
            OWSAssertIsOnMainThread();

            OWSLogInfo(@"unlock screen lock unexpectedly failed.");

            self.isShowingAuthUI = NO;

            // Local Authentication isn't working properly.
            // This isn't covered by the docs or the forums but in practice
            // it appears to be effective to retry again after waiting a bit.
            dispatch_async(dispatch_get_main_queue(), ^{
                [self ensureUI];
            });
        }
        cancel:^{
            OWSAssertIsOnMainThread();

            OWSLogInfo(@"unlock screen lock cancelled.");

            self.isShowingAuthUI = NO;

            [self ensureUI];
        }];

    [self ensureUI];
}

- (void)ensureUI
{
    [self updateUIWithState:ScreenLockUIStateScreenLock isLogoAtTop:NO animated:NO];
}

- (void)showScreenLockFailureAlertWithMessage:(NSString *)message
{
    OWSAssertIsOnMainThread();

    [OWSActionSheets showActionSheetWithTitle:OWSLocalizedString(@"SCREEN_LOCK_UNLOCK_FAILED",
                                                  @"Title for alert indicating that screen lock could not be unlocked.")
                                      message:message
                                  buttonTitle:nil
                                 buttonAction:^(ActionSheetAction *action) {
                                     // After the alert, update the UI.
                                     [self ensureUI];
                                 }
                           fromViewController:self];
}

- (void)dismissPressed:(id)sender
{
    OWSLogDebug(@"tapped dismiss share button");

    [self cancelShareExperience];
}

- (void)cancelShareExperience
{
    [self.shareViewDelegate shareViewWasCancelled];
}

#pragma mark - ScreenLockViewDelegate

- (void)unlockButtonWasTapped
{
    OWSAssertIsOnMainThread();

    OWSLogInfo(@"unlockButtonWasTapped");

    [self tryToPresentAuthUIToUnlockScreenLock];
}

@end

NS_ASSUME_NONNULL_END

File: SignalShareExtension/SAEScreenLockViewController.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalMessaging
import SignalUI

final class SAEScreenLockViewController: ScreenLockViewController {

    private weak var shareViewDelegate: ShareViewDelegate?

    init(shareViewDelegate: ShareViewDelegate) {
        self.shareViewDelegate = shareViewDelegate
        super.init(nibName: nil, bundle: nil)
        delegate = self
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - UIViewController

    override func loadView() {
        super.loadView()
        view.backgroundColor = Theme.launchScreenBackgroundColor
        title = OWSLocalizedString("SHARE_EXTENSION_VIEW_TITLE", comment: "Title for the 'share extension' view.")
        navigationItem.leftBarButtonItem = UIBarButtonItem(
            barButtonSystemItem: .stop,
            target: self,
            action: #selector(dismissPressed)
        )
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        ensureUI()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        ensureUI()

        if !hasShownAuthUIOnce {
            hasShownAuthUIOnce = true
            tryToPresentAuthUIToUnlockScreenLock()
        }
    }

    // MARK: - Lock Screen UI

    private var hasShownAuthUIOnce = false

    private var isShowingAuthUI = false

    private func ensureUI() {
        updateUIWithState(.screenLock, isLogoAtTop: false, animated: false)
    }

    private func tryToPresentAuthUIToUnlockScreenLock() {
        AssertIsOnMainThread()

        guard !isShowingAuthUI else { return }
        isShowingAuthUI = true

        ScreenLock.shared.tryToUnlockScreenLock(
            success: {
                AssertIsOnMainThread()

                Logger.info("unlock screen lock succeeded.")

                self.isShowingAuthUI = false
                self.shareViewDelegate?.shareViewWasUnlocked()
            },
            failure: { error in
                AssertIsOnMainThread()

                Logger.info("unlock screen lock failed.")

                self.isShowingAuthUI = false
                self.ensureUI()
                self.showScreenLockFailureAlertWithMessage(error.userErrorDescription)
            },
            unexpectedFailure: { error in
                AssertIsOnMainThread()

                Logger.info("unlock screen lock unexpectedly failed.")

                self.isShowingAuthUI = false

                // Local Authentication isn't working properly.
                // This isn't covered by the docs or the forums but in practice
                // it appears to be effective to retry again after waiting a bit.
                DispatchQueue.main.async {
                    self.ensureUI()
                }
            },
            cancel: {
                AssertIsOnMainThread()

                Logger.info("unlock screen lock cancelled.")

                self.isShowingAuthUI = false
                self.ensureUI()
            }
        )

        ensureUI()
    }

    private func showScreenLockFailureAlertWithMessage(_ message: String) {
        AssertIsOnMainThread()

        OWSActionSheets.showActionSheet(
            title: OWSLocalizedString(
                "SCREEN_LOCK_UNLOCK_FAILED",
                comment: "Title for alert indicating that screen lock could not be unlocked."
            ),
            buttonAction: { _ in
                // After the alert, update the UI.
                self.ensureUI()
            },
            fromViewController: self
        )
    }

    private func cancelShareExperience() {
        shareViewDelegate?.shareViewWasCancelled()
    }

    @objc
    private func dismissPressed(_ sender: Any) {
        Logger.debug("tapped dismiss share button")
        cancelShareExperience()
    }
}

extension SAEScreenLockViewController: ScreenLockViewDelegate {

    func unlockButtonWasTapped() {
        AssertIsOnMainThread()

        Logger.info("unlockButtonWasTapped")

        tryToPresentAuthUIToUnlockScreenLock()
    }
}
